### PR TITLE
Create Dynamic Template Based on PR Content

### LINKED ISSUE #87: PR checklist improvements
Looking to improve the PR checklist further, but not quite sure how. Ideally I'd want less text (or at least less up in the face) and more focus on anything the PR author writes. Perhaps split the template into various boxes.

Possibly add a script that adds labels based on what files are changed in the PR, e.g. whether they are content files or programming related.

In due course we'll try to shift more of the manual work, such as inserting metadata, over to Prepyrus.

### PR DESCRIPTION
**Pull request by JMielbrecht**
close #87 

## PR Author's Note

I've added a GitHub action to automatically label the PR based on the file types modified and edited the PR template for neatness and simplicity by eliminating checkboxes and organizing instructions for content/code contributions under the relevant dropdown. **_Note:_** The PR template below is outdated and the changes I've made to the template will only be visible once merged into `main`. In the meantime, here's some screenshots of what the checklist would be:

image

image

## Checklist

-   [ ] Philosophical or literary contribution (docs). Leave **unchecked** for
    <code>code</code> contribution.

    -   **IMPLIED CONSENT** By opening this pull request and contributing
        philosophical or literary content, I accept that my writing is submitted
        under the
        [ATTRIBUTION-NONCOMMERCIAL-SHAREALIKE 4.0 INTERNATIONAL](https://creativecommons.org/licenses/by-nc-sa/4.0/),
        which:

        -   Prohibits commercial reuse of the content.
        -   Allows sharing, remixing, and building upon the material as long as
            attribution is given.

        I understand that my writing may be modified, remixed, and built upon by
        others within the `systemphil/sphil` or sPhil project, in accordance
        with the license terms, indefinitely. See
        [legal code](https://creativecommons.org/licenses/by-nc-sa/4.0/legalcode.en).

    -   **REQUIRED** I have followed the
        [formatting guidelines](https://github.com/systemphil/sphil/blob/dev/src/pages/contributing/formatting/basic-markdown.md)
        and verified there are no formatting bugs.
        [Try markdown preview here](https://markdownlivepreview.com/).
    -   **REQUIRED** I have followed the
        [Chicago author-date style](https://www.chicagomanualofstyle.org/tools_citationguide/citation-guide-2.html).
    -   **REQUIRED** I have added or verified metadata title, description, and
        contributors at the _very top_ of the file followed by a `##` title
        heading. Additionally, I have ensured `isArticle` is set to `true`.
        Example:

        ```md
        ---
        title: The Immediate Difference Between Pure Being and Pure Nothing
        description:
            Learn about the difference between being and nothing in Hegel's
            Science of Logic.
        isArticle: true
        authors: Jerry Maguire (2024)
        editors: Steve Stevenson (2023), Karen Hansen (2022)
        contributors:
        ---

        ## My Article Title
        ```

        <details>

        <summary>Further information</summary>

        I have signed the document with my name/username under either as
        `Authors`, `Editors` or `Contributors`.

        > Use **Authors** if you have created and substantially added content.  
        > Use **Editor** if you have made substantial edits or review.  
        > Use **Contributor** if you have made minor edits, reviews or
        > contributions.  
        > If you've done multiple, pick the most weighted: Author > Editor >
        > Contributor.  
        > If you prefer to remain anonymous, that's fine too, but note that a
        > record of your contributions based on your GitHub username will exist
        > here in the codebase.

        </details>

    -   **REQUIRED** I have ensured that the
        [project's central bibliography](https://github.com/systemphil/sphil/blob/main/README_BIBLIOGRAPHY.md)
        contains the necessary bibliographical details for the citations I have
        used.

    -   _Optional_ My article is a stub or I want to actively encourage
        contribution, I've added the Stub component to the bottom of my content
        or where relevant:

        ```ts
        import Stub from "@/components/Stub";

        <Stub />;
        ```

-   **If Docs contribution is unchecked:** Code contribution
    ([Apache version 2 license](https://www.apache.org/licenses/LICENSE-2.0.txt))
      <details>

    All code apart of what is inside `src/pages/**` (excluding
    `/contributing/**`, `_app.mdx`, `_document.tsx`, `_meta.json`,
    `acknowledgements.mdx`, `index.mdx`, `privacy.mdx`, `team.mdx`, `terms.mdx`)
    is subject to Apache version 2 license. Basically, anything outside of
    content, literature, philosophy.

      </details>

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -1,98 +1,76 @@
...
+4. Add metadata at the top of your file:
+
+```md
+---
+title: Your Title Here
+description: Brief description of your content
+isArticle: true
+authors: Your Name (Year)
+editors: Editor Name (Year)
+contributors: Contributor Name
```
Comment (Firgrep) [inline .github/PULL_REQUEST_TEMPLATE.md:33]: You can add (Year) to this one as well

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -1,98 +1,76 @@
...
+Consider adding the Stub component to encourage further contributions.
+
+</details>
+
+<details>
+<summary>💻 Guidelines for Code Contributions</summary>
+
+### Required Steps
+
+1. Ensure code adheres to project coding standards
```
Comment (Firgrep) [inline .github/PULL_REQUEST_TEMPLATE.md:50]: Haha, this means we'll have to write what they are! Once things stabilize a bit, we'll update README and add a CONTRIBUTING textfile in the repo. Perhaps you could open an issue on this. :)

File: .github/workflows/pr-labeler.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+name: PR Labeler
+on: [pull_request]
+
+jobs:
+  label:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
```
Comment (Firgrep) [inline .github/workflows/pr-labeler.yml:8]: Outdated version. Latest is v4

File: .github/workflows/pr-labeler.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    - uses: actions/github-script@v6
```
Comment (Firgrep) [inline .github/workflows/pr-labeler.yml:9]: Outdated version. Latest is v7.

File: .github/workflows/pr-labeler.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      with:
+        github-token: ${{secrets.GITHUB_TOKEN}}
```
Comment (Firgrep) [inline .github/workflows/pr-labeler.yml:11]: The github token should be something made available via the permissions field at the top level, so I don't think it should be required, but I can be wrong here.

File: .github/workflows/pr-labeler.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        script: |
+          const fs = require('fs');
+          const path = require('path');
+
+          function getFileExtensions(files) {
+            return files.map(file => path.extname(file).toLowerCase()).filter(ext => ext !== '');
```
Comment (Firgrep) [inline .github/workflows/pr-labeler.yml:17]: Wouldn't this scan the entire codebase? What we want are the changes made in the PR, so we should extract that first?

File: .github/workflows/pr-labeler.yml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            const labels = [];
+            const CONTENT_EXTENSIONS = ['.mdx', '.md', '.svg', '.jpg', '.png'];
+            const CODE_EXTENSIONS = ['.tsx,', '.ts', '.html', '.css', '.json', '.toml', '.lock', '.yml', '.yaml', '.js'];
+        
+            if (extensions.some(ext => CONTENT_EXTENSIONS.includes(ext))) {
+              labels.push('content-contribution');
+            } else if (extensions.some(ext => CODE_EXTENSIONS.includes(ext))) {
+              labels.push('code-contribution'); 
+            } else {
+              labels.push('miscellaneous-contribution');
```
Comment (Firgrep) [inline .github/workflows/pr-labeler.yml:30]: This shouldn't be necessary. It's either content or code, or both, I reckon.

File: .github/workflows/pr-labeler.yml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+          const path = require('path');
+
+          function getFileExtensions(files) {
+            return files.map(file => path.extname(file).toLowerCase()).filter(ext => ext !== '');
+          }
+
+          async function addLabels(extensions) {
+            const labels = [];
+            const CONTENT_EXTENSIONS = ['.mdx', '.md', '.svg', '.jpg', '.png'];
+            const CODE_EXTENSIONS = ['.tsx,', '.ts', '.html', '.css', '.json', '.toml', '.lock', '.yml', '.yaml', '.js'];
```
Comment (Firgrep) [inline .github/workflows/pr-labeler.yml:23]: I don't think this is needed. We should only have to scan for any .mdx changes and add the content label based on that. Any other changes default to code.

File: .github/workflows/pr-labeler.yml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        
+            if (extensions.some(ext => CONTENT_EXTENSIONS.includes(ext))) {
+              labels.push('content-contribution');
+            } else if (extensions.some(ext => CODE_EXTENSIONS.includes(ext))) {
+              labels.push('code-contribution'); 
+            } else {
+              labels.push('miscellaneous-contribution'); 
+            }
+            
+            if (labels.length > 0) {
```
Comment (Firgrep) [inline .github/workflows/pr-labeler.yml:33]: This is looks like a good move! Bundle it together and do it in one request. 👍

**Review by Firgrep — CHANGES_REQUESTED** [review]
The PR template is lookig good! Some minor adjustments to that.

There are some bigger questions regarding the workflow. I reckon it work by granting permissions within the PR. 

The thing is, this is very hard to test without merging it in. Is it possible to test in the fork? Otherwise, would you consider setting up a quick dummy repo with a few files and folders to just test this workflow?

I got this thing autogenerated, could it be helpful? (see details below)

<details>

```
name: PR File Change Labeler

on:
  pull_request:
    types: [opened, synchronize, reopened]

permissions:
  contents: read
  pull-requests: write

jobs:
  label-pr:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

      - name: Label PR based on changed files
        uses: actions/github-script@v7
        with:
          script: |
            const { data: files } = await github.rest.pulls.listFiles({
              owner: context.repo.owner,
              repo: context.repo.repo,
              pull_number: context.issue.number
            });
            
            let hasContentChanges = false;
            let hasCodeChanges = false;
            
            // Check each changed file
            files.forEach(file => {
              if (file.filename.endsWith('.mdx')) {
                hasContentChanges = true;
              } else {
                hasCodeChanges = true;
              }
            });
            
            // Get current labels
            const { data: currentLabels } = await github.rest.issues.listLabelsOnIssue({
              owner: context.repo.owner,
              repo: context.repo.repo,
              issue_number: context.issue.number
            });
            
            const currentLabelNames = currentLabels.map(label => label.name);
            
            // Add labels if they don't already exist
            if (hasContentChanges && !currentLabelNames.includes('CONTENT')) {
              await github.rest.issues.addLabels({
                owner: context.repo.owner,
                repo: context.repo.repo,
                issue_number: context.issue.number,
                labels: ['CONTENT']
              });
            }
            
            if (hasCodeChanges && !currentLabelNames.includes('CODE')) {
              await github.rest.issues.addLabels({
                owner: context.repo.owner,
                repo: context.repo.repo,
                issue_number: context.issue.number,
                labels: ['CODE']
              });
            }
```

</details>

**Comment by JMielbrecht** [comment]
> The PR template is lookig good! Some minor adjustments to that.
> 
> 
> 
> There are some bigger questions regarding the workflow. I reckon it work by granting permissions within the PR. 
> 
> 
> 
> The thing is, this is very hard to test without merging it in. Is it possible to test in the fork? Otherwise, would you consider setting up a quick dummy repo with a few files and folders to just test this workflow?

Actually GitHub Actions can be tested in the PR itself. As you can see, the auto-label action has ran (although it encountered errors). If you go to my forked repository, I've opened a PR running a very similar script (although this one is based off of file paths and not file extensions) that works perfectly. You can check out that PR here for reference: https://github.com/JMielbrecht/sphil/pull/3

I'm not sure what you mean by “granting permissions within the PR”. If someone with admin permissions could create a personal access token with write permissions to PR’s, and add this token as a secret to the sPhil GitHub organization, I think this would solve the 403 error I’m getting in the workflow logs within this PR. I’m basing this off of the fact that the workflow in my aforementioned fork PR works perfectly fine, presumably because it’s my PR and I have admin access for that fork.  


I'll address your other feedback later tonight when I get off work. Thanks for the feedback!

File: .github/workflows/pr-labeler.yml (hunk 6 above)
Comment (JMielbrecht) [reply .github/workflows/pr-labeler.yml:17]: I believe on line 43, the GitHub REST API is being called but specifically for listing pull request files. I don't think it's scanning the entire codebase.

File: .github/workflows/pr-labeler.yml (hunk 7 above)
Comment (JMielbrecht) [reply .github/workflows/pr-labeler.yml:30]: Fair enough. I'll simplify the conditional

File: .github/workflows/pr-labeler.yml (hunk 8 above)
Comment (Firgrep) [reply .github/workflows/pr-labeler.yml]: Yeah, ok, keep the `.md` !

**Review by Firgrep — APPROVED** [review]

**Pull request merged**